app/lib/katello/event_daemon/monitor.rb

```diff
@@ -0,0 +1,50 @@
+module Katello
+  module EventDaemon
+    class Monitor
+      def initialize(services)
+        @services = services
+        @service_statuses = {}
+      end
+
+      def start
+        error = nil
+        status = nil
+        loop do
+          Rails.application.executor.wrap do
+            check_services(error, status)
+          end
+          sleep 15
+        end
+      end
+
+      def check_services(error, status)
+        @service_statuses = {}
+        @services.each do |service_name, service_class|
+          @service_statuses[service_name] = service_class.status
+        rescue => error
+          Rails.logger.error("Error occurred while pinging #{service_class}")
+          Rails.logger.error(error.message)
+          Rails.logger.error(error.backtrace.join("\n"))
+        ensure
+          if error || !@service_statuses[service_name]&.dig(:running)
+            begin
+              service_class.close
+              service_class.run
+              @service_statuses[service_name] = service_class.status
+            rescue => error
+              Rails.logger.error("Error occurred while starting #{service_class}")
+              Rails.logger.error(error.message)
+              Rails.logger.error(error.backtrace.join("\n"))
+            ensure
+              error = nil
+            end
+          end
+          Rails.cache.write(
```

Comment:
Let's add a 1 minute expiry to this cache key. That way we can be more sure of the accuracy in case, somehow, the monitor itself rails

Reply:
Seems that since we're now using `Rails.cache.read` instead of `fetch`, it will just return `nil` if the expiry time has passed.  Unlike `fetch`, it won't refresh anything.  So I'm not sure of the benefit of expiring it, unless we were to switch back to using `fetch`.

Reply:
> in case, somehow, the monitor itself rails

perhaps you meant ... goes off the rails? :smile:

Reply:
The underlying question here is "how long do we want this cache key to live?" Options include:

- briefly (seconds)
- short amount of time (minutes)
- a long time (hours or days)
- indefinitely (until the cache is cleared)

My suggestion was to prevent us from running into a situation where the Monitor is not running (who knows why), but the cached status (living until cache clear in the current implementation)  gives the impression that everything is OK. By giving it a reasonable expiry (1 minute) we can be sure that within the last minute that Monitor has set the status that we are looking at. I see it as a safety mechanism

Reply:
Hmm but maybe this is a bad idea cause then the ping API will show Not Running in between clearing the cache and the next tick of the monitor thread :(

Reply:
Another option would be to ditch the 15-second loop altogether and return to using `Rails.cache.fetch` with expiration.  That way when you Katello::Ping, it would either give you the (valid) cached data or update it on the spot.  But I feel like that's not really in scope of this PR.